feat: add share in twitter button to the end of project.

BalanaguYashwanth opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Ref #46178 of share button feature.

**Notes -**

Added share feature at end of the each project/course and along with implemented the custom twitter card

<!-- Feel free to add any additional description of changes below this line -->

BalanaguYashwanth committed:
added share button feature

BalanaguYashwanth committed:
updated share button function

BalanaguYashwanth committed:
updated test cases and imports suggested by codefactor

Sboonny reviewed (commented):
The issue didn't communicate what is needed, which cased you to do extra work, sorry about that should have noticed.

@ahmadabdolsaheb, can you direct the camper if we need share button that opens twitter embad card, or just copy the challenge.

I don't have specific feature in mind.

BalanaguYashwanth commented:
Hey @ahmadabdolsaheb, 

so there is no issue for raising PR for this share feature right

BalanaguYashwanth committed:
upgraded copied feature

BalanaguYashwanth committed:
merged with origin main to local branch

BalanaguYashwanth committed:
updated the tests

BalanaguYashwanth committed:
Merge branch 'freeCodeCamp:main' into feature/AddShareButtonEndOfProject

BalanaguYashwanth committed:
Merge remote-tracking branch 'origin/main' into feature/AddShareButtonEndOfProject

BalanaguYashwanth committed:
Merge branch 'feature/AddShareButtonEndOfProject' of https://github.com/BalanaguYashwanth/freeCodeCamp into feature/AddShareButtonEndOfProject

BalanaguYashwanth commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -184,32 +189,49 @@ const LowerJaw = ({
     return (
       <div>
         <hr />
-        <div className='lower-jaw-icon-bar'>
-          <button
-            className='btn fade-in'
-            title={t('buttons.reset-step')}
-            aria-label={t('buttons.reset-step')}
-            data-cy='reset-code-button'
-            onClick={openResetModal}
-          >
-            <Reset />
-          </button>
```
ok

BalanaguYashwanth commented on client/src/templates/Challenges/components/completion-modal.tsx:
```diff
@@ -314,6 +314,10 @@ class CompletionModalInner extends Component<
+          {currentBlockIds &&
+            currentBlockIds[currentBlockIds.length - 1] === id && (
+              <Share superBlock={superBlock} block={block} />
+            )}
```
ok

BalanaguYashwanth commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -184,32 +189,49 @@ const LowerJaw = ({
             </button>
           ) : null}
+
+          {challengeIsCompleted && endOfProject.current ? (
```
thanks for suggesting

BalanaguYashwanth commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -184,32 +189,49 @@ const LowerJaw = ({
+      data?.challengeNode.challenge.id
+    ) {
+      endOfProject.current = true;
+    }
```
I saw it is mentioned in the stories, to mention button at end of the project

BalanaguYashwanth committed:
updated shareComponent with twitterCard and unit tests

BalanaguYashwanth commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -184,32 +189,49 @@ const LowerJaw = ({
-          </button>
-
+        <div>
           {isAttemptsLargerThanTest && !challengeIsCompleted ? (
```
I didn't understand about your comment @Sboonny, Two code looks same that you suggested and which i had wrote

Sboonny replied:
empty line for format

BalanaguYashwanth commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -47,6 +51,7 @@ const LowerJaw = ({
   const { t } = useTranslation();
   const submitButtonRef = React.createRef<HTMLButtonElement>();
   const testFeedbackRef = React.createRef<HTMLDivElement>();
+  const endOfProject = React.useRef<boolean>(false);
```
I understand @Sboonny, 

As per @ahmadabdolsaheb user stories we need to show **share button** at end of the each project.

If possible once discuss with @ahmadabdolsaheb and let me know
<img width="1015" alt="image" src="https://user-images.githubusercontent.com/36238382/224524354-90777c05-608d-4f9e-bc47-2d606f36607b.png">

Sboonny replied:
I will open new issue after you are done, you don't need to stop your progress for this

BalanaguYashwanth replied:
ok

BalanaguYashwanth committed:
Merge remote-tracking branch 'origin/main' into feature/AddShareButtonEndOfProject

BalanaguYashwanth commented:
Hey guys,

Today after taking latest i see you shifted to pnpm then after i ran **pnpm lint**. 

From existing codebase i saw so many lint issues, can you give me some suggestion how to clear them ?
<img width="1014" alt="image" src="https://user-images.githubusercontent.com/36238382/224525120-b99d406b-fa3e-4e3b-a7de-075453eb9194.png">

Sboonny commented:
If you need help head to [our chat room](https://discord.gg/PRyKn3Vbay) or [the forum](https://forum.freecodecamp.org/c/contributors/3) if you need help contributing, our prolific contributors will guide you through this.

Because most of the time, you find the answer quicker there, and you may not know if the maintainers are on break because of the weekends or not.

---

What commend did you use?

I have run Gitpod, and I didn't have an issue. Try `pnpm install` and see if it still around

BalanaguYashwanth committed:
updated shareComponent redirect link  and unit tests

BalanaguYashwanth committed:
updated unit tests for share template component

BalanaguYashwanth commented:
Just now committed, but i didn't faced any lint issues, but here few checks are not passing. Any idea ?

<img width="962" alt="image" src="https://user-images.githubusercontent.com/36238382/224527137-2fa29094-f604-48db-8270-919fb3e8c3a0.png">

BalanaguYashwanth committed:
formatted code

BalanaguYashwanth committed:
updated the snapshot of share template

BalanaguYashwanth committed:
updated the snapshot of share template

BalanaguYashwanth committed:
updated unit tests of share template

BalanaguYashwanth commented:
Hey guys,

I think most of the things are completed can you review once

ahmadabdolsaheb commented:
@BalanaguYashwanth, thanks for working on implementing the twitter card feature. I had a quick look and here are a few recomendations:
- You could connect the lower jaw component to redux and pull  superBlock, completedPercent. That way passing around data and creating getChallengesList would not be necessary.
- "copied-succesfully": "Copied succesfully!!!", is still in the translations.json ( I believe it was for the pervious 
- I completed a challenge on the /2022/responsive-web-design/ but the button did not show up. Am I missing something?

BalanaguYashwanth commented:
Hello @ahmadabdolsaheb,

> I completed a challenge on the /2022/responsive-web-design/ but the button did not show up. Am I missing something?

If you are talking about twitter button, then it will appear when you complete the last module of these expanded drop downs. 

In this case, if you complete 69th then twitter button will show
Reference - 
<img width="751" alt="image" src="https://user-images.githubusercontent.com/36238382/224736261-29d8af54-e321-4527-a70a-fac00a105950.png">

BalanaguYashwanth commented:
> "copied-succesfully": "Copied succesfully!!!", is still in the translations.json ( I believe it was for the pervious

Yes, Will remove it, thanks for letting me

ahmadabdolsaheb commented:
> In this case, if you complete 69th then twitter button will show

It makes sense to show the button only at the end of the block; however, if a user does not fully complete the challenge, they will have an incomplete percentage on their card.

For instance, if they only complete the 69th challenge, the card would say "I have completed 1٪ of Learn HTML by Building a Cat Photo App #freecodecamp"

If you are planning to show the button at the block completion ( I think it is a great idea ), you should display it when the completion percentage is 100%.   (creating a isBlockCompletedSelector is preferred) 

That way the percentage can be dropped completely from the card and it can say:
"I  have completed Learn HTML by Building a Cat Photo App #freecodecamp"

This would help improve the translations as well.

BalanaguYashwanth commented:
Sure, i will doing changes that your interested

BalanaguYashwanth commented:
Also when i was connecting to **lowerJaw component to pull superBlock, completedPercent**, I am facing error like lowerJaw should be inside provider store then when i kept inside provider, 

I am not getting proper values instead i am getting initial store values like ``complete percent always showing 0``

when i see remaining components like editor they are not wrapped inside provider, i am not able to getting how they wrapped inside the provider store ?

ahmadabdolsaheb commented:
push your changes and let us have a look.

BalanaguYashwanth committed:
Updated with twitter card

BalanaguYashwanth committed:
Merge remote-tracking branch 'origin/main' into feature/AddShareButtonEndOfProject

BalanaguYashwanth committed:
Updated with filenames as per lint suggestions

Sboonny commented on client/i18n/locales/english/translations.json:
```diff
@@ -517,7 +517,7 @@
   },
   "flash": {
-    "honest-first": "To claim a certification, you must first agree to our academic honesty policy",
+    "honest-first": "To claim a certification, you must first accept our academic honesty policy",
```
```suggestion
    "honest-first": "To claim a certification, you must first agree to our academic honesty policy",
```

ahmadabdolsaheb replied:
This change is out of scope. Please revert the changes or commit the suggestion.

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -197,7 +218,14 @@ const LowerJaw = ({
             <Reset />
           </button>
-
+          {challengeIsCompleted &&
+            endOfProject.current &&
+            completedPercent === isBlockCompleted && (
```
Can you turn this if statements into a variable for readiblity, something like 

```ts
const showShareButton = challengeIsCompleted &&
            endOfProject.current &&
            completedPercent === isBlockCompleted

     {showShareButton ** (
```

BalanaguYashwanth replied:
updated thanks for suggesting

Sboonny reviewed (commented):
I haven't given in depth review yet, but these suggestions are cleaning old changes to make it easier to review it later

BalanaguYashwanth committed:
Updated code with given comments and twitter card

BalanaguYashwanth committed:
Removed unncessary code

Sboonny commented on api-server/src/server/utils/get-curriculum.js:
```diff
@@ -19,3 +19,17 @@ export function getChallenges() {
     }, []);
 }
+
+export function getChallengesList({ userInputSuperBlock, userInputBlock }) {
+  for (let superBlock in curriculum) {
+    if (superBlock === userInputSuperBlock) {
+      for (let block in curriculum[superBlock].blocks) {
+        if (block === userInputBlock) {
+          return curriculum[superBlock].blocks[block].challenges.sort(
+            (a, b) => a.title.split(' ')[1] - b.title.split(' ')[1]
+          );
+        }
+      }
+    }
+  }
+}
```
@ahmadabdolsaheb, we don't need this logic if your PR #49479 is merged, right? If all we need to know that they have completed 100% of the challenges, we can import the value from ProgressPar 🤔, and we don't need the API calls.

 correct me if I am misunderstanding something

BalanaguYashwanth replied:
ok, Once PR https://github.com/freeCodeCamp/freeCodeCamp/pull/49479 is merged, let me know once. I will update the code.

ahmadabdolsaheb replied:
https://github.com/freeCodeCamp/freeCodeCamp/pull/49479 has been merged.

Sboonny commented on client/src/components/share/index.tsx:
```diff
@@ -0,0 +1,16 @@
+  block: string;
+}
+
+export const Share: React.FC<ShareProps> = ({ superBlock, block }) => {
```
This isn't Functional component, as there isn't `Share.something`, and using it has its own downside, you can find more here https://github.com/facebook/create-react-app/pull/8177

```suggestion
export const Share = ({ superBlock, block }: ShareProps): JSX.Element => {
```

BalanaguYashwanth replied:
Updating with this change

Sboonny commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
```
Can you move `share-template` into `index.tsx`, there isn't a complicate pieces in it, and moving them away provide hazard more than comfort.

BalanaguYashwanth replied:
I can move to index.tsx, By doing that you are compromising on TDD unit tests and scalability of code, 

If i merged to index.tsx. then i can't write proper unit test cases, it redirects to code scalability reduced. if we start extend this more features in future.

Sboonny replied:
True, but the issue doesn't require any additional feature to it 👍, I will leave the discussion to @ahmadabdolsaheb, because he knows the roadmap better.

ahmadabdolsaheb replied:
There are pros and cons for both scenarios. We can keep it as is for now.

Sboonny commented on client/src/components/share/use-share.tsx:
```diff
@@ -0,0 +1,36 @@
+import i18next from 'i18next';
```
Can you move `use-share` hook into `index.tsx`, there isn't a complicate pieces in it, and moving them away provide hazard more than comfort.

BalanaguYashwanth replied:
I can move to index.tsx, By doing that you are compromising on TDD unit tests and scalability of code, 

If i merged to index.tsx. then i can't write proper unit test cases, it redirects to code scalability reduced. if we start extend this more features in future.

Sboonny replied:
```suggestion
import { useTranslation } from "react-i18next";
```

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -166,7 +177,7 @@ const LowerJaw = ({
     }
   };
-
+  const isBlockCompleted = 100;
```
This should be a global variable, so it doesn't render everytime the component is mounted

BalanaguYashwanth replied:
ok updating

Sboonny reviewed (commented):
Awesome work @BalanaguYashwanth, it looks nice, here some nitpicks

BalanaguYashwanth committed:
updated few code changes for twitter share component based on code reviewers

ahmadabdolsaheb commented:
> Hello @ahmadabdolsaheb ,
> 
> > You could connect the lower jaw component to redux and pull superBlock, completedPercent.
> 
> I will bring this change after this #49479 PR get merged, so that i can get these params values like superBlock, completedPercent with updated lowerjaw code

sounds like a plan

Sboonny commented:
@BalanaguYashwanth, Ahmad pr is merged 👍

BalanaguYashwanth commented:
ok will look into it

BalanaguYashwanth committed:
Merge 'origin/main' to feature/AddShare.. branch

BalanaguYashwanth commented:
Hello @ahmadabdolsaheb,

As you said i am trying to connect the redux for ``lower-jaw.tsx`` , To get the values of superBlock, block and completedPercentage.

So, I saw some examples👇 how to call those values using with redux ``connect``, 



<img width="430" alt="image" src="https://user-images.githubusercontent.com/36238382/229200304-24af1fcb-9989-478b-8ef5-d9ef53b46d58.png">

I was applied same to the ``lower-jaw.tsx``, 
Initially it asked me to wrap inside the provider and pass store, Then after wrapping with provider, Later then i getting empty values of these superBlock, block and completedPercent.

So am i doing any wrong, Do I need to connect ``lower-jaw.tsx`` to the global redux provider ?

BalanaguYashwanth commented:
Hello @ahmadabdolsaheb,

After analyzing, **how we can connect ``lower-jaw.tsx`` to redux**,  

- First, I saw that ``lower-jaw.tsx`` is not connected context.Provider and provider

To achieve this to connect with redux.

1) Either we need to push the ``lower-jaw.tsx`` to provider section where all components are coming under the hood

(or)

2) We need to initialize the context.Provider and provider again into the `lower-jaw.tsx`` to get the values like superBlock and block.

<img width="649" alt="image" src="https://user-images.githubusercontent.com/36238382/229268176-54a10769-328d-45e9-affa-d6d5e0923cd1.png">

So, Need your inputs -
1) which is the best solution from above two of them ?.
2) Check is it neccessary to connect redux to `lower-jaw.tsx`` or not ?.

ahmadabdolsaheb commented:
Thanks for digging into the issue.  To evaluate our options better, It seems we could isolate connecting the lowerjaw to redux into an issue/pr of its own.
For now, feel free to proceed by pulling the required states from redux using the parent component's (editor) mapStateToProps and passing them to the lowerJaw component.

BalanaguYashwanth commented:
ok awesome, let me update everything and push it everthing

BalanaguYashwanth committed:
removed unnecessary imports and getting info from editor

BalanaguYashwanth committed:
Merge remote-tracking branch 'origin/main' into feature/AddShareButtonEndOfProject

BalanaguYashwanth commented:
Everything updated

BalanaguYashwanth commented:
Everything updated, @ahmadabdolsaheb you can check now

 CC: @Sboonny

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -165,6 +186,18 @@ const LowerJaw = ({
     );
   };
 
+  const challengeOrder: ChallengeWithCompletedNode[] = getChallengesList({
+    userInputSuperBlock: superBlock || '',
+    userInputBlock: block || ''
+  }) as ChallengeWithCompletedNode[];
+  if (challengeOrder[challengeOrder.length - 1].id === lastChallengeId) {
+    endOfProject.current = true;
+  }
+  const showShareButton =
+    challengeIsCompleted &&
+    endOfProject.current &&
+    completedPercent === isBlockCompleted;
```
```suggestion
  const showShareButton =
    challengeIsCompleted && completedPercent === isBlockCompleted;
```

BalanaguYashwanth replied:
Hello @Sboonny,

Mainly this share button need to show only at **end of Project** as per requirement 👇. So added the ``endOfProject.current`` logic.

If i remove this logic then if you achieved 100% and if you went to the start or middle of the project then this share button will show. but as per @ahmadabdolsaheb requirement, we need only at the end of the project

I understand if user goes unordered way to complete the challenges then it make sense showing share button at any point of the project, 

Discuss with your team (@ahmadabdolsaheb) about requirement and update the requirement then will remove the logic

<img width="825" alt="image" src="https://user-images.githubusercontent.com/36238382/230704428-7c47a60e-909e-4fbf-aaf1-88f7d4a3b669.png">

ahmadabdolsaheb replied:
Updated the spec to read "Add a share button below the progress bar at the project completion".

BalanaguYashwanth replied:
Thanks for your response, will update soon

BalanaguYashwanth replied:
Hey really sorry for not updating, getting some PNPM issues, will try to update in 1 or 2 days. Won't make it much late

Sboonny replied:
Take your time, there is no hurry. if you face an issue, tell us in discord or the forum.

BalanaguYashwanth replied:
Hello @Sboonny ,

I updated it code, also some other unit tests are failing to update snapshot, i am updating in local but again those are failing in build, please check on it once

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -150,6 +169,8 @@ const LowerJaw = ({
   const [isFeedbackHidden, setIsFeedbackHidden] = useState(false);
   const { t } = useTranslation();
   const testFeedbackRef = React.createRef<HTMLDivElement>();
+  const endOfProject = React.useRef<boolean>(false);
+  const isBlockCompleted = 100;
```
```suggestion
  const { id: superBlock, block } = challengeMeta;
```

BalanaguYashwanth replied:
I am destructing it, when taking props itself into function. so i think it what you suggested for destructing and my code destructing looks almost similar. 

Still if you want me to improve i can improve it

<img width="524" alt="image" src="https://user-images.githubusercontent.com/36238382/230704653-80952243-c4ee-492b-ab41-837c04e63334.png">

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -132,6 +149,8 @@ const LowerJawStatus = ({
 };
 
 const LowerJaw = ({
```
```suggestion
const isBlockCompleted = 100;

const LowerJaw = ({
```

BalanaguYashwanth replied:
Hey thanks updated this

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -18,6 +24,9 @@ interface LowerJawPanelProps {
   helpButtonEvent: () => void;
   hideHelpButton: boolean;
+  showShareButton: boolean;
+  superBlock: string;
+  block: string;
```
The type can be cleaned if you import the shareProps

here is the syntax

```ts
interface LowerJawPanelProps extends ShareProps
```

```suggestion
```

BalanaguYashwanth replied:
I didn't tried this way, and this is very good suggestion

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -132,6 +149,8 @@ const LowerJawStatus = ({
 };
 
 const LowerJaw = ({
+  challengeMeta: { id: lastChallengeId, superBlock, block },
```
```suggestion
  challengeMeta,
```

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -35,6 +44,8 @@ interface LowerJawStatusProps {
 }
 
 interface LowerJawProps {
```
```suggestion
export interface LowerJawProps {
```

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -9,6 +9,12 @@ import Help from '../../../assets/icons/help';
 import Reset from '../../../assets/icons/reset';
 import { MAX_MOBILE_WIDTH } from '../../../../../config/misc';
 import { apiLocation } from '../../../../../config/env.json';
+import {
+  ChallengeMeta,
+  ChallengeWithCompletedNode
+} from '../../../redux/prop-types';
+import { getChallengesList } from '../../../../../api-server/src/server/utils/get-curriculum';
```
```suggestion
import { ChallengeMeta } from '../../../redux/prop-types';
import type { ShareProps } from '../../../components/share/use-share';
```

Sboonny commented on client/src/templates/Challenges/classic/editor.tsx:
```diff
@@ -68,6 +70,8 @@ import './editor.css';
 const MonacoEditor = Loadable(() => import('react-monaco-editor'));
 
 export interface EditorProps {
```
You need to import the `LowerJawProps`

```suggestion
export interface EditorProps extends Pick<LowerJawProps, 'challengeMeta' | 'completedPercent'> {
```

Sboonny commented on client/src/templates/Challenges/classic/editor.tsx:
```diff
@@ -68,6 +70,8 @@ import './editor.css';
 const MonacoEditor = Loadable(() => import('react-monaco-editor'));
 
 export interface EditorProps {
+  challengeMeta: ChallengeMeta;
+  completedPercent: number;
```
```suggestion
```

Sboonny commented on client/src/components/share/use-share.tsx:
```diff
@@ -0,0 +1,36 @@
+import i18next from 'i18next';
+
+interface ShareTemplateProps {
+  superBlock: string;
+  block: string;
+}
```
ShareProps is the same as breadCrumbProps, can you link both, to reduce the repeated types.

```suggestion
export type ShareProps = BreadCrumbProps
```

Sboonny commented on client/src/components/share/use-share.tsx:
```diff
@@ -0,0 +1,36 @@
+export const useShare = ({
+  superBlock,
+  block
+}: ShareTemplateProps): ShareTemplateReturnProps => {
```
The new type

```suggestion
}: ShareProps): ShareRedirectProps=> {
```

BalanaguYashwanth replied:
done

Sboonny commented on client/src/components/share/use-share.tsx:
```diff
@@ -0,0 +1,36 @@
+  superBlock,
+  block
+}: ShareTemplateProps): ShareTemplateReturnProps => {
+  const redirectFreeCodeCampLearnURL = `https://${freecodecampLearnDomainURL}/${superBlock}/${hastag}${block}`;
+  const i18nSupportedBlock =
+    i18next.t(`intro:${superBlock}.blocks.${block}.title`) || block;
```
no need to import everything from `i18next` for `t`, you can use `useTranslation` hook, and there is no need to fallback to `block`, if both of them always true value. 

```suggestion
  const { t } = useTranslation();
  const redirectFreeCodeCampLearnURL = `https://${freecodecampLearnDomainURL}/${superBlock}/${hastag}${block}`;
  const i18nSupportedBlock = t(`intro:${superBlock}.blocks.${block}.title`);

```

Sboonny replied:
Note that if both aren't true, then the type should be changed accordingly

BalanaguYashwanth replied:
ok

Sboonny commented on client/src/components/share/use-share.tsx:
```diff
@@ -0,0 +1,36 @@
+  block: string;
+}
+
+interface ShareTemplateReturnProps {
```
```suggestion
export interface ShareRedirectProps{
```

Sboonny commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import { faTwitter } from '@fortawesome/free-brands-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+
+export interface ShareTemplateProps {
+  handleRedirectToTwitter: () => void;
+}
+
+export const ShareTemplate: React.FC<ShareTemplateProps> = ({
+  handleRedirectToTwitter
+}) => {
```
```suggestion
import type { ShareRedirectProps } from './use-share';

export const ShareTemplate = ({
  handleRedirectToTwitter
}: Pick<ShareRedirectProps, 'handleRedirectToTwitter'>): JSX.Element => {
```

BalanaguYashwanth replied:
🤯Awesome i didn't knew this.

Sboonny commented on client/src/components/share/index.tsx:
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { ShareTemplate } from './share-template';
+import { useShare } from './use-share';
+
+interface ShareProps {
+  superBlock: string;
+  block: string;
+}
```
```suggestion
import { useShare, type ShareProps } from './use-share';
```

BalanaguYashwanth replied:
I removed ``ShareProps `` from ``'./use-share'`` and kept in index.tsx as the single point of ``ShareProps`` interface.

Hope it gets reusing everywhere

Sboonny reviewed (commented):
Before Review can you address the previous suggestions and here are new suggestion that:

- Cleaning the extra logic
- Cleaning the extra types

---

And please handle the failing test, it's failing in both firefox and chrome, if you think it's a flaky test, reopen and close the PR.

BalanaguYashwanth committed:
Cleaned the code and resued the interfaces

BalanaguYashwanth commented:
I didn't expected these many comments, and atlast updated those comments

BalanaguYashwanth committed:
updated unit tests for useshare

ahmadabdolsaheb commented:
Thank you for making progress on this feature. The related PRs have been merged. If you could make the changes according to the proposed suggestions, we could go for the final round of reviews.

BalanaguYashwanth committed:
Merge branch 'freeCodeCamp:main' into feature/AddShareButtonEndOfProject

bbsmooth commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,18 @@
+      className='btn fade-in'
+      onClick={handleRedirectToTwitter}
+    >
+      <FontAwesomeIcon title='twitterIcon' icon={faTwitter} size='1x' />
```
```suggestion
      <FontAwesomeIcon icon={faTwitter} size='1x' aria-hidden="true"/>
```

The icon should be hidden from screen reader users. We can use either sr-only text or `aria-label` to provide an accessible name for this which screen reader users will hear.

BalanaguYashwanth replied:
Thanks for keeping your time for adding suggestions

bbsmooth commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,18 @@
+  handleRedirectToTwitter
+}: Pick<ShareRedirectProps, 'handleRedirectToTwitter'>): JSX.Element => {
+  return (
+    <button
```
This should be a link not a button since it takes you to Twitter's web page. It also needs an accessible name added using either sr-only text or `aria-label`, probably something like "Share on Twitter". You can also add a `title` attribute with the same string you use for the accessible name.

Sboonny replied:
it should look like a button, not be a button

naomi-lgbt replied:
Yes, it should be a link, but you can style the link to look like a button.

BalanaguYashwanth replied:
ok i will try to do like that

bbsmooth replied:
Yes, for accessibility. The general rule is "links take you places, buttons do things.". When a screen reader user hears that it is a link then they will know that when they activate the link they will go to a different page. Also, since this links opens a new window we need to add "opens in new window" onto the end of the name so screen reader users will be prepared for this as well. It would be a little strange to have that wording for a button.

BalanaguYashwanth replied:
Updated

BalanaguYashwanth committed:
updated changes as per PR comments removing fn and added accessibility

BalanaguYashwanth committed:
Merge remote-tracking branch 'origin/main' into feature/AddShareButtonEndOfProject

BalanaguYashwanth committed:
updated link by removing button for twitter share

BalanaguYashwanth committed:
added text for link

naomi-lgbt reviewed (dismissed):
The snapshots are now out of sync with the codebase - you'll need to run `pnpm run test-client -u` to update the snapshots, and commit + push the changes.

BalanaguYashwanth committed:
Update client/src/templates/Challenges/classic/editor.tsx

I can't remove this becauselower-jaw.tsx using this params, also As @ahmadabdolsaheb said we can't connect lower-jaw.tsx to redux.

Altogether will. create separate issue for connecting lower-jaw.tsx to redux.

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

BalanaguYashwanth committed:
Update client/i18n/locales/english/translations.json

ok not adding this change

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

BalanaguYashwanth committed:
updated interface props as per PR comments

BalanaguYashwanth committed:
Merge branch 'feature/AddShareButtonEndOfProject' of https://github.com/BalanaguYashwanth/freeCodeCamp into feature/AddShareButtonEndOfProject

BalanaguYashwanth committed:
Merge branch 'main' of https://github.com/BalanaguYashwanth/freeCodeCamp into feature/AddShareButtonEndOfProject

bbsmooth commented:
Just a thought. Since we are adding text to the reset and help buttons (PR #50087) should we also add text to this twitter link?

bbsmooth commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,30 @@
+      target='_blank'
+      rel='noreferrer'
+    >
+      <span className='sr-only'>opens in new window</span>
+      <span className='sr-only'>{t('share-on-twitter')}</span>
```
```suggestion
      <span className='sr-only'>{t('buttons.share-on-twitter')}, {t('aria.opens-new-window')}</span>
```

"Opens in new window" should come after the link text and it already has a translation we should use.

Also, this suggestion is based on how the link currently looks. If we decide to display "Twitter" next to the link then this will change a little.

BalanaguYashwanth replied:
Actually this text is surrounded by link

bbsmooth replied:
@BalanaguYashwanth I'm not sure what your last comment is trying to say. When a screen reader hears this link it should say the name of the button first and then "opens in new window". That's the change I suggested.

BalanaguYashwanth replied:
ok got it now

bbsmooth commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,30 @@
+  const { t } = useTranslation();
+  return (
+    <a
+      title='Share on Twitter'
```
```suggestion
      title={t('buttons.share-on-twitter') || 'Share on Twitter'}
```

Need to use the translation here. Fallback added to appease TypeScript.

Sboonny commented:
> should we also add text to this twitter link?

For me, it depends on which merge first.

BalanaguYashwanth committed:
updated share template text according to screen reader

BalanaguYashwanth committed:
Merge branch 'main' of https://github.com/BalanaguYashwanth/freeCodeCamp into feature/AddShareButtonEndOfProject

BalanaguYashwanth committed:
updated test case

BalanaguYashwanth committed:
removed bad snapshot

BalanaguYashwanth committed:
removed bad snapshot txt

BalanaguYashwanth committed:
connected lowerjaw to redux

ahmaxed committed:
Merge branch main into feature/AddShareButtonEndOfProject

ahmaxed committed:
fix: clean up

ahmaxed committed:
feat: adjust the look for the tweet button

naomi-lgbt reviewed (approved):
Some food for thought. Because this only appears in the lower jaw, campers won't see it when they complete a certification project.

I wonder if we want to show it in the completion modal too? Does not necessarily have to be in this PR, but might be nice. 

<img width="675" alt="image" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/63889819/1cdf11b4-5f2e-4c6b-8f8e-4c936b8e72f2">

Otherwise, the tweet flow itself looks good: 

<img width="478" alt="image" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/63889819/3ab4853a-78c5-4e5f-9fad-a6a342a797f9">

Sboonny commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -69,6 +93,9 @@ const LowerButtonsPanel = ({
           <Reset />
           {resetButtonText}
         </button>
+        {showShareButton && (
+          <Share superBlock={superBlock || ''} block={block || ''} />
+        )}
```
```suggestion
        {showShareButton && <Share superBlock={superBlock} block={block} />}
```

Sboonny committed:
Clean unneed if statements

Sboonny reviewed (approved):
#### :sparkles: :tada: **AWESOME!** :tada: :sparkles:

 Hi @BalanaguYashwanth, thanks for this pull request and for contributing to the code-base for the first time. We are looking forward to more contributions from you in the future.

 Cheers & happy contributing!

BalanaguYashwanth commented:
🎉Really Amazing, I saw that @ahmaxed and @Sboonny you guys kept lot of efforts than me😭.
 
I understand that i need to take lot more responsibility of my tasks which i pickup, So that you guys no need to keep much efforts to give multiple feedbacks and followups to iterate.

Its a good experience I learned lot of things, Tried my level best to improve it. Next time onwards i will keep more responsibility.

bbsmooth commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,32 @@
+      target='_blank'
+      rel='noreferrer'
+    >
+      <span className='sr-only'>
+        {t('buttons.share-on-twitter')}, {t('aria.opens-new-window')}
+      </span>
```
```suggestion
```

See below

BalanaguYashwanth replied:
Updating

BalanaguYashwanth replied:
updated

bbsmooth commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,32 @@
+        aria-label='twitterIcon'
+        aria-hidden='true'
+      />
+      <span aria-hidden>{t('buttons.tweet')}</span>
```
```suggestion
      {t('buttons.tweet')} <span className='sr-only'>
        {t('aria.opens-new-window')}
      </span>
```

If the visual text of this button is just going to be "Tweet" then the accessible name should be the same (except we also need to add "Opens in new window").

BalanaguYashwanth replied:
updated

BalanaguYashwanth replied:
tested it working good after adding this change

bbsmooth commented on client/src/components/share/share-template.tsx:
```diff
@@ -0,0 +1,32 @@
+  const { t } = useTranslation();
+  return (
+    <a
+      title={t('buttons.share-on-twitter') || 'Share on Twitter'}
```
I'm not completely opposed to using the `title` attribute here so that the browser will show this as a tooltip, but keyboard users won't see it. Do we really feel this is essential information? Is the word "Tweet" on the button not enough? If so, then perhaps we should figure out a way to give everyone this necessary information.

BalanaguYashwanth replied:
Need your suggestions @ahmaxed @Sboonny 😅

Sboonny replied:
I think the word "tweet" is descriptive, we can delete the attribute and wait for feedback, if we need it to be more descriptive.

```suggestion
```

BalanaguYashwanth committed:
updated changes as per accessibility

BalanaguYashwanth committed:
sync changes from main

BalanaguYashwanth committed:
updated tests

ahmaxed committed:
Update client/src/components/share/share-template.tsx

Co-authored-by: Muhammed Mustafa <MuhammedElruby@gmail.com>

ahmaxed commented:
@naomi-lgbt, thanks for the review. Yes, the scope of this PR was limited intentionally. 
Here are a few follow up PRs to improve the feature:
1. Add e2e tests since #51014 is merged
2. Add a button for the projects
3. Improve the twitter card to include custom image 

@BalanaguYashwanth, thank you for updating the pr multiple times as spec changed. I look forward to your future contributions. Since you are the domain expert, feel free to tackle any of the above mentioned tasks by opening related issues.

Sboonny commented on client/i18n/locales/english/translations.json:
```diff
@@ -77,6 +77,8 @@
     "change-language": "Change Language",
     "resume-project": "Resume project",
     "start-project": "Start project",
+    "share-on-twitter": "Share on Twitter",
```
```suggestion
```

Sboonny committed:
Remove extra string

ahmaxed commented:
@Sboonny, I think there are three additional files added in this commit: https://github.com/freeCodeCamp/freeCodeCamp/pull/49395/commits/6483fcb3a06c72beffe582ffa2bbdaee2faad28b

Sboonny commented:
solved in another PR

ahmaxed commented:
Cool. Thanks Oliver
